archive/frames/views.py

```diff
@@ -422,6 +425,81 @@ def get_endpoint_name(self):
                           'zip': 'getZipArchive'}
 
         return endpoint_names.get(self.action)
+    
+
+class ThumbnailViewSet(viewsets.ModelViewSet):
+    permission_classes = (AdminOrReadOnly,)
+    filter_backends = (
+        DjangoFilterBackend,
+    )
+    filter_class = ThumbnailFilter
+
+    def get_queryset(self):
+        """
+        Filter thumbnails depending on the logged in user.
+        Admin users see all thumbnails
+        Authenticated users see all frames with a PUBDAT in the past, plus
+        all thumbnails that belong to frames under their proposals.
+        Non authenticated see all thumbnails for frames with a PUBDAT in the past
+        """
+        queryset = (
+            Thumbnail.objects.all().select_related('frame')
+        )
+        if self.request.user.is_superuser:
+            return queryset
+        elif self.request.user.is_authenticated:
+            return queryset.filter(
+                Q(frame__proposal_id__in=self.request.user.profile.proposals) |
+                Q(frame__public_date__lt=datetime.datetime.now(datetime.timezone.utc))
+            )
+        else:
+            return queryset.filter(frame__public_date__lt=datetime.datetime.now(datetime.timezone.utc))
+
+    # These two method overrides just force the use of the as_dict method for serialization for list and detail endpoints
+    def list(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
+        page = self.paginate_queryset(queryset)
+        json_models = [model.as_dict() for model in page]
+        return self.get_paginated_response(json_models)
+
+    def retrieve(self, request, *args, **kwargs):
+        instance = self.get_object()
+        return Response(instance.as_dict())
+
+    def create(self, request):
+        basename = request.data.get('basename')
+        logger_tags = {'tags': {
+            'filename': basename,
+            'request_id': request.data.get('request_id')
+        }}
+        logger.info('Got request to process thumbnail', extra=logger_tags)
+        # Make sure we have the minimum information to make a frame object associated with the thumbnail if one doesn't already exist
+        frame_serializer = FrameSerializer(data=request.data)
+        if frame_serializer.is_valid():
```

Comment:
This bit is probably the most complicated part of this. I've designed this so that we always expect a valid frame payload from the ingester so we're able to create a basic Frame object. I've done this to keep the changes in the ingester minimal - the ingester still uploads to s3 and gets the version information, but rather than storing the key and extension on the frame, we grab it and store it on the thumbnail instead, and store the Frame without that info.